build-common/src/org/jetbrains/kotlin/incremental/storage/RelocatableFileToPathConverter.kt

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.incremental.storage
+
+import java.io.File
+
+/**
+ * [FileToPathConverter] which converts a [File] to a Unix-style relative path, which is relative to a base directory ([baseDir]), and vice
+ * versa.
+ *
+ * This is to support build cache relocatability
+ * (https://docs.gradle.org/current/userguide/build_cache_concepts.html#relocatability).
+ */
+class RelocatableFileToPathConverter(private val baseDir: File) : FileToPathConverter {
+
+    private val unixStyleBaseDirPathPrefix = "${baseDir.invariantSeparatorsPath}/"
+
+    override fun toPath(file: File): String {
+        check(file.invariantSeparatorsPath.startsWith(unixStyleBaseDirPathPrefix)) {
+            "The given file '${file.path}' is located outside the base directory '${baseDir.path}'"
+        }
```

Comment:
Can compiler plugins  output .class files directly? If so, can they specify a custom location and will those classes still be processed by the IC caches?

Reply:
If the compiler plugins generate `.class` files outside `classesDir`, the IC will not be aware of those (there's currently no API for the two to communicate on this info), so those classes will not be processed by the IC caches.

Reply:
Ok, then this sounds like a safe assertion.